### DISCUSSION TITLE
Filtering Select options

### DISCUSSION THREAD

**Discussion by yaplej** [filamentphp/filament#2531] [Help] [answered]
I have looked through the Demo site code to find how to filter options for a Select field based on the current records related field but was not finding anything.  

How would I get this to work?  I keep getting `Argument #1 ($record) must be of type App\Models\Record, Illuminate\Database\Query\Builder given`

I am only trying to present options that match the Record's RecordType.

```
Forms\Components\Select::make('recordoption_id')
                    ->options(RecordOption::where('recordtype_id', fn (Record $record) => $record->recordtype_id)->pluck('name', 'id'))
                    ->searchable()
                    ->label('Record Option'),
```

**Comment by ryangjchandler** [comment, marked as answer]
@yaplej You want to provide a callback function to the `options()` method instead:

```php
->options(function (Closure $get) {
	return RecordOption::where('recordtype_id', $get('recordtype_id'))->pluck('name', 'id');
})
```

**Reply by yaplej** [reply]
That is returning no results.  I think that the callback is returning the current Record id.  Not the recordtype_id of the Record.

I am working with Record and trying to pull in all the valid RecordOptions.

**Reply by ryangjchandler** [reply]
Try something like this:

```php
->options(function (Record $record) {
	return RecordOption::where('recordtype_id', $record->recordtype_id)->pluck('name', 'id');
})
```

Be careful though, since this will likely fail on the create form of your resource because `$record` will be `null`.

**Reply by yaplej** [reply]
Its still not populating.  I can use ::all and it works.  Only when trying to limit/filter the options it is empty.

```
->options(function (Record $record) {
	return RecordOption::all()->pluck('name', 'id');
})
```

**Reply by ryangjchandler** [reply]
@yaplej Have you tried debugging / dumping out `$record` to see what the property's value is?

**Reply by yaplej** [reply]
I have not.  I also ran into what you were warning about.  This method totally breaks creating new records. :(

I guess the proper way to do this would be a responsive method?  Not sure what that looks like.

**Reply by yaplej** [reply]
So to dynamically generate ->options() based on the form field changes it looks like I need to use Closure customizations?
https://filamentphp.com/docs/2.x/forms/advanced#using-closure-customisation

I could store the `recordtype_id` as a variable.  This example in the documentation does not cover doing this in the Admin ModelResource though its for Forms so maybe things are a little different?

```
use Closure;
 
function (Closure $get) {
    $recordtype_id= $get('recordtype_id'); // Store the value of the `recordtype_id` field in the `$recordtype_id` variable.
    //...
}
```

Then maybe I could get the Closure method you recommended to work?
```
->options(function (Closure $get) {
	return RecordOption::where('recordtype_id', $get('recordtype_id'))->pluck('name', 'id');
})
```

Is their an example of doing this in the Demo already?

**Reply by yaplej** [reply]
I just wanted to come back and say that this did work just as you recommended.  I don't know what my problem was but maybe I was missing  `use Closure;` or something the first time I tried it.
